message/msg_addr.hh

```diff
@@ -10,18 +10,25 @@
 
 #include "gms/inet_address.hh"
 #include <cstdint>
+#include "locator/host_id.hh"
 
 namespace netw {
 
 struct msg_addr {
     gms::inet_address addr;
     uint32_t cpu_id;
-    friend bool operator==(const msg_addr& x, const msg_addr& y) noexcept;
-    friend bool operator<(const msg_addr& x, const msg_addr& y) noexcept;
-    friend std::ostream& operator<<(std::ostream& os, const msg_addr& x);
+    locator::host_id host_id;
+
+    msg_addr(gms::inet_address ip, uint32_t cpu = 0, locator::host_id host_id = locator::host_id::create_null_id()) noexcept
+        : addr(ip)
+        , cpu_id(cpu)
+        , host_id(host_id)
```

Comment:
Can "ip" be optional? That is, can one create msg_addr with just the host_id set? Part of https://github.com/scylladb/scylladb/issues/6403 was to allow ip-less messaging.

Reply:
Until the transition is done, I don't think we can always get the ip address from the host_id in all paths, in particular in the gossiper, so I'm afraid we still need to send the ip address.
I suggest that the next step, after the transition is complete, should be to ignore the ip address and even reset it on the destination side to prove it's not needed, and only then it'd be safe to drop it on the client side.

Reply:
We may still need to have ip-based addressing, but we should make clear what are the semantics of msg_addr in the messaging layer. For example, if I have msg_addr with both host id and the ip set, is the message guaranteed to be delivered to the given host id? This is not clarified in this patch. 

I don't think we want semantics like "try to deliver to host id, but if not found, deliver to the IP", because that kills the safety of addressing by host id, and algorithms are not guaranteed to be correct.

We could make msg_addr hold variant<inet_address, host_id>.